// react-mobile-picker/dist/react-mobile-picker.js
import me, { createContext as re, useMemo as k, useReducer as pe, useCallback as g, useContext as ne, useState as q, useEffect as ee, useRef as J } from "react";
var te = { exports: {} }, W = {};
/**
 * @license React
 * react-jsx-runtime.production.js
 *
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var le;
function he() {
  if (le)
    return W;
  le = 1;
  var s = Symbol.for("react.transitional.element"), c = Symbol.for("react.fragment");
  function r(i, a, n) {
    var l = null;
    if (n !== void 0 && (l = "" + n), a.key !== void 0 && (l = "" + a.key), "key" in a) {
      n = {};
      for (var f in a)
        f !== "key" && (n[f] = a[f]);
    } else
      n = a;
    return a = n.ref, {
      $$typeof: s,
      type: i,
      key: l,
      ref: a !== void 0 ? a : null,
      props: n
    };
  }
  return W.Fragment = c, W.jsx = r, W.jsxs = r, W;
}
var G = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var ue;
function Ee() {
  return ue || (ue = 1, process.env.NODE_ENV !== "production" && function() {
    function s(e) {
      if (e == null)
        return null;
      if (typeof e == "function")
        return e.$$typeof === X ? null : e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case R:
          return "Fragment";
        case b:
          return "Profiler";
        case j:
          return "StrictMode";
        case _:
          return "Suspense";
        case $:
          return "SuspenseList";
        case I:
          return "Activity";
      }
      if (typeof e == "object")
        switch (typeof e.tag == "number" && console.error(
          "Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."
        ), e.$$typeof) {
          case h:
            return "Portal";
          case w:
            return (e.displayName || "Context") + ".Provider";
          case v:
            return (e._context.displayName || "Context") + ".Consumer";
          case Y:
            var t = e.render;
            return e = e.displayName, e || (e = t.displayName || t.name || "", e = e !== "" ? "ForwardRef(" + e + ")" : "ForwardRef"), e;
          case H:
            return t = e.displayName || null, t !== null ? t : s(e.type) || "Memo";
          case N:
            t = e._payload, e = e._init;
            try {
              return s(e(t));
            } catch {
            }
        }
      return null;
    }
    function c(e) {
      return "" + e;
    }
    function r(e) {
      try {
        c(e);
        var t = !1;
      } catch {
        t = !0;
      }
      if (t) {
        t = console;
        var o = t.error, u = typeof Symbol == "function" && Symbol.toStringTag && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return o.call(
          t,
          "The provided key is an unsupported type %s. This value must be coerced to a string before using it here.",
          u
        ), c(e);
      }
    }
    function i(e) {
      if (e === R)
        return "<>";
      if (typeof e == "object" && e !== null && e.$$typeof === N)
        return "<...>";
      try {
        var t = s(e);
        return t ? "<" + t + ">" : "<...>";
      } catch {
        return "<...>";
      }
    }
    function a() {
      var e = y.A;
      return e === null ? null : e.getOwner();
    }
    function n() {
      return Error("react-stack-top-frame");
    }
    function l(e) {
      if (z.call(e, "key")) {
        var t = Object.getOwnPropertyDescriptor(e, "key").get;
        if (t && t.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function f(e, t) {
      function o() {
        F || (F = !0, console.error(
          "%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://react.dev/link/special-props)",
          t
        ));
      }
      o.isReactWarning = !0, Object.defineProperty(e, "key", {
        get: o,
        configurable: !0
      });
    }
    function E() {
      var e = s(this.type);
      return L[e] || (L[e] = !0, console.error(
        "Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release."
      )), e = this.props.ref, e !== void 0 ? e : null;
    }
    function O(e, t, o, u, A, P, Z, Q) {
      return o = P.ref, e = {
        $$typeof: T,
        type: e,
        key: t,
        props: P,
        _owner: A
      }, (o !== void 0 ? o : null) !== null ? Object.defineProperty(e, "ref", {
        enumerable: !1,
        get: E
      }) : Object.defineProperty(e, "ref", { enumerable: !1, value: null }), e._store = {}, Object.defineProperty(e._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: 0
      }), Object.defineProperty(e, "_debugInfo", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: null
      }), Object.defineProperty(e, "_debugStack", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: Z
      }), Object.defineProperty(e, "_debugTask", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: Q
      }), Object.freeze && (Object.freeze(e.props), Object.freeze(e)), e;
    }
    function m(e, t, o, u, A, P, Z, Q) {
      var d = t.children;
      if (d !== void 0)
        if (u)
          if (B(d)) {
            for (u = 0; u < d.length; u++)
              S(d[u]);
            Object.freeze && Object.freeze(d);
          } else
            console.error(
              "React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead."
            );
        else
          S(d);
      if (z.call(t, "key")) {
        d = s(e);
        var M = Object.keys(t).filter(function(de) {
          return de !== "key";
        });
        u = 0 < M.length ? "{key: someKey, " + M.join(": ..., ") + ": ...}" : "{key: someKey}", V[d + u] || (M = 0 < M.length ? "{" + M.join(": ..., ") + ": ...}" : "{}", console.error(
          `A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`,
          u,
          d,
          M,
          d
        ), V[d + u] = !0);
      }
      if (d = null, o !== void 0 && (r(o), d = "" + o), l(t) && (r(t.key), d = "" + t.key), "key" in t) {
        o = {};
        for (var K in t)
          K !== "key" && (o[K] = t[K]);
      } else
        o = t;
      return d && f(
        o,
        typeof e == "function" ? e.displayName || e.name || "Unknown" : e
      ), O(
        e,
        d,
        P,
        A,
        a(),
        o,
        Z,
        Q
      );
    }
    function S(e) {
      typeof e == "object" && e !== null && e.$$typeof === T && e._store && (e._store.validated = 1);
    }
    var p = me, T = Symbol.for("react.transitional.element"), h = Symbol.for("react.portal"), R = Symbol.for("react.fragment"), j = Symbol.for("react.strict_mode"), b = Symbol.for("react.profiler"), v = Symbol.for("react.consumer"), w = Symbol.for("react.context"), Y = Symbol.for("react.forward_ref"), _ = Symbol.for("react.suspense"), $ = Symbol.for("react.suspense_list"), H = Symbol.for("react.memo"), N = Symbol.for("react.lazy"), I = Symbol.for("react.activity"), X = Symbol.for("react.client.reference"), y = p.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE, z = Object.prototype.hasOwnProperty, B = Array.isArray, C = console.createTask ? console.createTask : function() {
      return null;
    };
    p = {
      "react-stack-bottom-frame": function(e) {
        return e();
      }
    };
    var F, L = {}, D = p["react-stack-bottom-frame"].bind(
      p,
      n
    )(), U = C(i(n)), V = {};
    G.Fragment = R, G.jsx = function(e, t, o, u, A) {
      var P = 1e4 > y.recentlyCreatedOwnerStacks++;
      return m(
        e,
        t,
        o,
        !1,
        u,
        A,
        P ? Error("react-stack-top-frame") : D,
        P ? C(i(e)) : U
      );
    }, G.jsxs = function(e, t, o, u, A) {
      var P = 1e4 > y.recentlyCreatedOwnerStacks++;
      return m(
        e,
        t,
        o,
        !0,
        u,
        A,
        P ? Error("react-stack-top-frame") : D,
        P ? C(i(e)) : U
      );
    };
  }()), G;
}
process.env.NODE_ENV === "production" ? te.exports = he() : te.exports = Ee();
var x = te.exports;
const Te = 216, ve = 36, ge = "off", oe = re(null);
oe.displayName = "PickerDataContext";
function ae(s) {
  const c = ne(oe);
  if (c === null) {
    const r = new Error(`<${s} /> is missing a parent <Picker /> component.`);
    throw Error.captureStackTrace && Error.captureStackTrace(r, ae), r;
  }
  return c;
}
const ce = re(null);
ce.displayName = "PickerActionsContext";
function se(s) {
  const c = ne(ce);
  if (c === null) {
    const r = new Error(`<${s} /> is missing a parent <Picker /> component.`);
    throw Error.captureStackTrace && Error.captureStackTrace(r, se), r;
  }
  return c;
}
function ke(s, c = (r) => r) {
  return s.slice().sort((r, i) => {
    const a = c(r), n = c(i);
    if (a === null || n === null)
      return 0;
    const l = a.compareDocumentPosition(n);
    return l & Node.DOCUMENT_POSITION_FOLLOWING ? -1 : l & Node.DOCUMENT_POSITION_PRECEDING ? 1 : 0;
  });
}
function Re(s, c) {
  switch (c.type) {
    case "REGISTER_OPTION": {
      const { key: r, option: i } = c;
      let a = [...s[r] || [], i];
      return a = ke(a, (n) => n.element.current), {
        ...s,
        [r]: a
      };
    }
    case "UNREGISTER_OPTION": {
      const { key: r, option: i } = c;
      return {
        ...s,
        [r]: (s[r] || []).filter((a) => a !== i)
      };
    }
    default:
      throw Error(`Unknown action: ${c.type}`);
  }
}
function be(s) {
  const {
    style: c,
    children: r,
    value: i,
    onChange: a,
    height: n = Te,
    itemHeight: l = ve,
    wheelMode: f = ge,
    ...E
  } = s, O = k(
    () => ({
      height: l,
      marginTop: -(l / 2),
      position: "absolute",
      top: "50%",
      left: 0,
      width: "100%",
      pointerEvents: "none"
    }),
    [l]
  ), m = k(
    () => ({
      height: `${n}px`,
      position: "relative",
      display: "flex",
      justifyContent: "center",
      overflow: "hidden",
      maskImage: "linear-gradient(to top, transparent, transparent 5%, white 20%, white 80%, transparent 95%, transparent)",
      WebkitMaskImage: "linear-gradient(to top, transparent, transparent 5%, white 20%, white 80%, transparent 95%, transparent)"
    }),
    [n]
  ), [S, p] = pe(Re, {}), T = k(
    () => ({ height: n, itemHeight: l, wheelMode: f, value: i, optionGroups: S }),
    [n, l, i, S, f]
  ), h = g((b, v) => {
    if (i[b] === v)
      return !1;
    const w = { ...i, [b]: v };
    return a(w, b), !0;
  }, [a, i]), R = g((b, v) => (p({ type: "REGISTER_OPTION", key: b, option: v }), () => p({ type: "UNREGISTER_OPTION", key: b, option: v })), []), j = k(
    () => ({ registerOption: R, change: h }),
    [R, h]
  );
  return /* @__PURE__ */ x.jsxs(
    "div",
    {
      style: {
        ...m,
        ...c
      },
      ...E,
      children: [
        /* @__PURE__ */ x.jsx(ce.Provider, { value: j, children: /* @__PURE__ */ x.jsx(oe.Provider, { value: T, children: r }) }),
        /* @__PURE__ */ x.jsxs(
          "div",
          {
            style: O,
            children: [
              /* @__PURE__ */ x.jsx(
                "div",
                {
                  style: {
                    position: "absolute",
                    top: 0,
                    bottom: "auto",
                    left: 0,
                    right: "auto",
                    width: "100%",
                    height: "1px",
                    background: "#d9d9d9",
                    transform: "scaleY(0.5)"
                  }
                }
              ),
              /* @__PURE__ */ x.jsx(
                "div",
                {
                  style: {
                    position: "absolute",
                    top: "auto",
                    bottom: 0,
                    left: 0,
                    right: "auto",
                    width: "100%",
                    height: "1px",
                    background: "#d9d9d9",
                    transform: "scaleY(0.5)"
                  }
                }
              )
            ]
          }
        )
      ]
    }
  );
}
const ie = re(null);
ie.displayName = "PickerColumnDataContext";
function fe(s) {
  const c = ne(ie);
  if (c === null) {
    const r = new Error(`<${s} /> is missing a parent <Picker.Column /> component.`);
    throw Error.captureStackTrace && Error.captureStackTrace(r, fe), r;
  }
  return c;
}
function _e({
  style: s,
  children: c,
  name: r,
  ...i
}) {
  const { height: a, itemHeight: n, wheelMode: l, value: f, optionGroups: E } = ae("Picker.Column"), O = k(
    () => f[r],
    [f, r]
  ), m = k(
    () => E[r] || [],
    [r, E]
  ), S = k(
    () => {
      let t = m.findIndex((o) => o.value === O);
      return t < 0 && (t = 0), t;
    },
    [m, O]
  ), p = k(
    () => a / 2 - n * m.length + n / 2,
    [a, n, m]
  ), T = k(
    () => a / 2 - n / 2,
    [a, n]
  ), [h, R] = q(0);
  ee(() => {
    R(a / 2 - n / 2 - S * n);
  }, [a, n, S]);
  const j = se("Picker.Column"), b = J(h);
  b.current = h;
  const v = g(() => {
    let t = 0;
    const o = b.current;
    o >= T ? t = 0 : o <= p ? t = m.length - 1 : t = -Math.round((o - T) / n), j.change(r, m[t].value) || R(a / 2 - n / 2 - t * n);
  }, [j, a, n, r, T, p, m]), [w, Y] = q(0), [_, $] = q(!1), [H, N] = q(0), I = g((t) => {
    t < p ? t = p - Math.pow(p - t, 0.8) : t > T && (t = T + Math.pow(t - T, 0.8)), R(t);
  }, [T, p]), X = g((t) => {
    N(t.targetTouches[0].pageY), Y(h);
  }, [h]), y = g((t) => {
    t.cancelable && t.preventDefault(), _ || $(!0);
    const o = w + t.targetTouches[0].pageY - H;
    I(o);
  }, [_, w, H, I]), z = g(() => {
    _ && ($(!1), N(0), Y(0), v());
  }, [v, _]), B = g(() => {
    _ && ($(!1), N(0), R(w), Y(0));
  }, [_, w]), C = J(null), F = g((t) => {
    if (t.deltaY === 0)
      return;
    let o = t.deltaY * 0.1;
    Math.abs(o) < n && (o = n * Math.sign(o)), l === "normal" && (o = -o);
    const u = h + o;
    I(u);
  }, [n, h, I, l]), L = g(() => {
    v();
  }, [v]), D = g((t) => {
    l !== "off" && (t.cancelable && t.preventDefault(), F(t), C.current && clearTimeout(C.current), C.current = setTimeout(() => {
      L();
    }, 200));
  }, [L, F, C, l]), U = J(null);
  ee(() => {
    const t = U.current;
    return t && (t.addEventListener("touchmove", y, { passive: !1 }), t.addEventListener("wheel", D, { passive: !1 })), () => {
      t && (t.removeEventListener("touchmove", y), t.removeEventListener("wheel", D));
    };
  }, [y, D]);
  const V = k(
    () => ({
      flex: "1 1 0%",
      maxHeight: "100%",
      transitionProperty: "transform",
      transitionTimingFunction: "cubic-bezier(0, 0, 0.2, 1)",
      transitionDuration: _ ? "0ms" : "300ms",
      transform: `translate3d(0, ${h}px, 0)`
    }),
    [h, _]
  ), e = k(
    () => ({ key: r }),
    [r]
  );
  return /* @__PURE__ */ x.jsx(
    "div",
    {
      style: {
        ...V,
        ...s
      },
      ref: U,
      onTouchStart: X,
      onTouchEnd: z,
      onTouchCancel: B,
      ...i,
      children: /* @__PURE__ */ x.jsx(ie.Provider, { value: e, children: c })
    }
  );
}
function Pe(s) {
  return typeof s == "function";
}
function xe({
  style: s,
  children: c,
  value: r,
  ...i
}) {
  const a = J(null), { itemHeight: n, value: l } = ae("Picker.Item"), f = se("Picker.Item"), { key: E } = fe("Picker.Item");
  ee(
    () => f.registerOption(E, { value: r, element: a }),
    [E, f, r]
  );
  const O = k(
    () => ({
      height: `${n}px`,
      display: "flex",
      justifyContent: "center",
      alignItems: "center"
    }),
    [n]
  ), m = g(() => {
    f.change(E, r);
  }, [f, E, r]);
  return /* @__PURE__ */ x.jsx(
    "div",
    {
      style: {
        ...O,
        ...s
      },
      ref: a,
      onClick: m,
      ...i,
      children: Pe(c) ? c({ selected: l[E] === r }) : c
    }
  );
}
const Se = Object.assign(be, {
  Column: _e,
  Item: xe
});
export {
  Se as default
};
